### ISSUE TITLE
Move test data from Julia code to JSON files

### ISSUE THREAD

**Issue by nhniina** [spine-tools/SpineOpt.jl#636]
Currently, test data is directly included in the Julia files in the test folder. It would be better to move the data into JSON files that are read by the tests. JSON files can be easily imported/exported and thus modified using the DB Editor.

How many JSON files should there be? Each test set could be a scenario in one big JSON file or each test set could be stored in its own JSON file. Maybe store at least unit test data and forthcoming system test data in separate JSON files?

**Comment by nhniina** [comment]
@jkiviluo , @LouisFliche

**Comment by DillonJ** [comment]
@manuelma what do you think? I can see advantages and disadvantages. Are there implementation issues? I remember that we use an in-memory database so the tests run faster, does that have any bearing on this?

Another issue is that in many of the test sets, the base system is modified programatically in the actual tests themselves, i.e. parameters and units added, for example. If we also have these in JSON files, it would be quite cumbersome when writing new tests. At the same time, we probably want the data all in the once place. I personally like that all the data is self contained in the tests themselves and you don't have to look outside. You see precisely the impact of adding new things incrementally we way they are written currently. They are also nice examples of how to manipulate Spine data using the API in Julia.

An advantage of having the data in JSON files could also be that we could use the data for system tests and perhaps for tutorials.

Also, what is best practice here?

**Comment by manuelma** [comment]
Reading and parsing a JSON file for each test will make the suite run slower, that's for sure. How much slower it's difficult to say. Maybe that's the only real disadvantage I can think of.

Apart from that, there's the issue of convenience. When I first read this proposal, I didn't feel it was a very good idea to be honest; but then again I thought it's really a matter of preference. Being able to import a test system into DB editor might be helpful to some developers. In my case, I prefer to just look at the code because the structure is so simply presented there (or maybe I'm used to it).

@DillonJ raises an important point, that many tests share a good deal of the data and thus the shared part can be defined only once in code (and the final test system is imported in two steps, first the shared data and then some additional data particular to each test). If we put the data in JSON with aims to manually import it into DB editor, I think we would like to import it in a single step. So this means readjusting the data a bit.

> 
@manuelma what do you think? I can see advantages and disadvantages. Are there implementation issues? I remember that we use an in-memory database so the tests run faster, does that have any bearing on this?

Another issue is that in many of the test sets, the base system is modified programatically in the actual tests themselves, i.e. parameters and units added, for example. If we also have these in JSON files, it would be quite cumbersome when writing new tests. At the same time, we probably want the data all in the once place. I personally like that all the data is self contained in the tests themselves and you don't have to look outside. You see precisely the impact of adding new things incrementally the way they are written currently. They are also nice examples of how to manipulate Spine data using the API in Julia.

An advantage of having the data in JSON files could also be that we could use the data for system tests and perhaps for tutorials.

> Also, what is best practice here?

Good point. We should look at other packages probably, like JuMP.

**Comment by manuelma** [comment]
An alternative approach would be to refactor the test package a little bit so the user is able to run each test individually by calling a function, say `test_constraint_unit_capacity`. We are already half way there I think.

But the new addition would be that the user can specify the url a real database, so the test is run by importing the data into that database (at the moment, as @DillonJ points above, the data is imported into an in-memory DB). This way the user could call the function, pass the url of their DB, and then open the DB in DB editor to see the data used in the test.

**Comment by DillonJ** [comment]
A hazard I can see of attempting to import a single shared dataset common to all or many test sets is the interdependency that is then created... one might wish to change the data for one test and inadvertently impact other tests. Right now the tests are pretty self contained.

A big advantage of using JSON data may be for system tests where the DBs would be bigger and much more difficult to create in code - I wonder should that be the focus? Unit tests are testing very specific things.

**Comment by jkiviluo** [comment]
The aim is to have datasets that serve multiple tests and are easier to maintain as a database rather than lines of code. It's mainly coming from the need to support 'system tests' in addition to 'unit tests'. This will be extended to the 'open certification process' at a later stage (and at that point it cannot be SpineOpt code as it should be in format that can serve multiple models). 

Using JSON rather than Spine DB as the storage format is just because of git - line changes will be visible and don't cause bloating of the repository as binary file changes would.

If the tests become noticeably slower, and there is no reasonable way to go mitigate that, maybe 'unit tests' could remain in code and only 'system tests' would be in JSON. It might also be good to have execution separation between them - 'unit tests' should be executed more often (automatically) while 'system tests' only when someone touches the interdependent parts of the code. @manuelma Could GitHub actions be set to do only part of the tests automatically?

BTW, does export of DB to JSON work consistently? I.e. do the lines change place randomly if exporting the same DB or is there e.g. an alphabetic ordering?

> ...one might wish to change the data for one test and inadvertently impact other tests.

That's a good point. I think most of the time new tests should just add data - not modify existing data. If someone modifies existing data, they would then have the responsibility to ensure that all tests continue working. I suppose that would be visible from the tests.

**Comment by manuelma** [comment]
I like the idea of keeping unit-tests the way they are, to simplify the already painful process of developing them.

And then system tests can be developed using JSON files.

> @manuelma Could GitHub actions be set to do only part of the tests automatically?

Absolutely.

> BTW, does export of DB to JSON work consistently? I.e. do the lines change place randomly if exporting the same DB or is there e.g. an alphabetic ordering?

I am 99% sure that's the case already but if not, it is very straightforward to do.

**Comment by jkiviluo** [comment]
Ok, let's keep unit tests in code for now at least and revisit only if those test start to feel practical on the system test side (which will be done in JSON).

Also, let's continue with the assumption that system tests will be triggered only manually.

**Comment by clizbe** [comment]
@nhniina So is the outcome to do something or we decided against it? (Maybe this can be closed or moved to a discussion.)
